out_s3: added sequential indexing feature to s3_key_format

StephenLeeY opened the pull request:
Signed-off-by: Stephen Lee <sleemamz@amazon.com>

<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->
https://github.com/fluent/fluent-bit-docs/pull/555

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

Specify `$INDEX` in `s3_key_format` to add an index that increments every time it uploads a
file. Sequential indexing is compatible with all other `s3_key_format features`, so `$UUID`
and `$TAG` can be specified at the same time as `$INDEX`. When `$INDEX` is specified but `$UUID`
is not, a random string will not be appended to the end of the key name.

If FluentBit crashes or fails to upload, the index will be preserved in the filesystem
in a metadata file located in `${store_dir}/index_metadata` and reused on startup. This
configuration option uses native C functions so is incompatible with multi-threading.

Tested through unit testing and various input plugins (exec, random, etc) as well as Valgrind.

### Example Configuration File
```
[INPUT]
    name exec
    command date +"%Y-%m-%d %H:%M:%S,%3N"

[OUTPUT]
    name s3
    match *
    region us-west-2
    bucket bucket-name
    s3_key_format /test/log-file-$INDEX.gz
    use_put_object true
    total_file_size 2M
    upload_timeout 10s
    compression gzip
    store_dir /tmp/fluent-bit/s3-output-buffer
    retry_limit 5
    log_key exec
```

### Output file names
```
...
log-file-3.gz
log-file-2.gz
log-file-1.gz
log-file-0.gz
```

## Valgrind
For some reason, when running with the example configuration file, the Valgrind is throwing many errors even on a clean master branch. For this reason, I believe it is not my code throwing these errors but the current master branch code. Below are the Valgrind logs for the clean master branch.
#### Clean Master Branch Valgrind Output Logs
```
==2337== HEAP SUMMARY:
==2337==     in use at exit: 701,635 bytes in 5,374 blocks
==2337==   total heap usage: 119,718 allocs, 114,344 frees, 13,844,281 bytes allocated
==2337==
==2337== LEAK SUMMARY:
==2337==    definitely lost: 0 bytes in 0 blocks
==2337==    indirectly lost: 0 bytes in 0 blocks
==2337==      possibly lost: 0 bytes in 0 blocks
==2337==    still reachable: 701,635 bytes in 5,374 blocks
==2337==         suppressed: 0 bytes in 0 blocks
==2337== Reachable blocks (those to which a pointer was found) are not shown.
==2337== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==2337==
==2337== For counts of detected and suppressed errors, rerun with: -v
==2337== ERROR SUMMARY: 16 errors from 11 contexts (suppressed: 0 from 0)
```
and here are the Valgrind logs for the example configuration file.
#### Example Configuration File Valgrind Logs
```
==5198== HEAP SUMMARY:
==5198==     in use at exit: 701,635 bytes in 5,374 blocks
==5198==   total heap usage: 119,741 allocs, 114,367 frees, 13,925,751 bytes allocated
==5198==
==5198== LEAK SUMMARY:
==5198==    definitely lost: 0 bytes in 0 blocks
==5198==    indirectly lost: 0 bytes in 0 blocks
==5198==      possibly lost: 0 bytes in 0 blocks
==5198==    still reachable: 701,635 bytes in 5,374 blocks
==5198==         suppressed: 0 bytes in 0 blocks
==5198== Reachable blocks (those to which a pointer was found) are not shown.
==5198== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==5198==
==5198== For counts of detected and suppressed errors, rerun with: -v
==5198== ERROR SUMMARY: 16 errors from 11 contexts (suppressed: 0 from 0)
```
They appear to be exactly the same.

StephenLeeY commented:
@PettitWesley

PettitWesley commented:
@StephenLeeY Please break this down into a nice set of commits, one for each component you modified, something like:
* out_s3: add...
* aws_util: ...
* tests: internal: ....

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -252,6 +253,58 @@ static flb_sds_t concat_path(char *p1, char *p2)
     return dir;
 }
 
+static int read_seq_index(char *seq_index_file, uint64_t *seq_index,
```
can you add a comment explaining how this function is used- why does it get two ints from the file?

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -252,6 +253,58 @@ static flb_sds_t concat_path(char *p1, char *p2)
+    }
+    *seq_index = ret;
+
+    ret = getw(fp);
```
So this worries me slightly because `getw` reads an amount of bytes from the file which is different on different systems. Its not platform independent.

But I think it is acceptable to use this function. I think for Fluent Bit we only care about 32 bit and 64 bit systems. On both, an int should be 4 bytes. 

Its important to always remember with C code- what assumptions am I making about what sort of system this code will run on?

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -423,6 +485,67 @@ static int cb_s3_init(struct flb_output_instance *ins,
+
+            /* Construct directories and file path names */
+            tmp = fs_stream->path;
+            ctx->metadata_dir = flb_sds_create(tmp);
```
for all of these calls to flb_sds_create and flb_sds_cat- they can fail to alloc in which case you need to check the return value and then call `flb_errno()` if it failed.

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -423,6 +485,67 @@ static int cb_s3_init(struct flb_output_instance *ins,
+
+            ctx->seq_index_file = flb_sds_create(ctx->metadata_dir);
+            tmp = "/seq_index_";
+            ctx->seq_index_file = flb_sds_cat(ctx->seq_index_file, tmp, strlen(tmp));
+            tmp = (char *) &ctx->ins->name;
+            ctx->seq_index_file = flb_sds_cat(ctx->seq_index_file, tmp, strlen(tmp));
```
Why concatenate the plugin name to the file name?

PettitWesley replied:
The plugin name should usually be something like `s3.0` or `s3.1`. However a user can give it their own name using an alias. I have no clue what the allowed char-set is for aliases. Are you certain that whatever this name is- it can be a file name? I wonder if the string needs to be sanitized?

StephenLeeY replied:
The purpose of concatenating is to prevent a scenario where multiple different S3 output plugins that are initialized in the same FluentBit configuration file have to share one metadata file.

For the plugin name, that is a good point. Sanitizing the name seems like a lot of unnecessary work, so I will just concatenate the id instead, and that should be good enough.

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -423,6 +485,67 @@ static int cb_s3_init(struct flb_output_instance *ins,
             flb_plg_error(ctx->ins, "'s3_key_format' must start with a '/'");
             return -1;
         }
+        if (strstr((char *) tmp, "$INDEX")) {
```
this is a huge amount of new and complex code for the init function- please move it to a helper function to make things easier to read/cleaner

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -423,6 +485,67 @@ static int cb_s3_init(struct flb_output_instance *ins,
+            } else {
+                ret = read_seq_index(ctx->seq_index_file, &ctx->seq_index, &ctx->decrement_index);
+                if (ret < 0) {
+                    flb_plg_error(ctx->ins, "Failed to read from metadata file");
```
this error message and the one above could be better, lets explicitly say sequence file. When users complain of problems running fluent bit... its good to have very descriptive and unique error messages to help them & us get an idea of what went wrong

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -1113,6 +1240,19 @@ static int s3_put_object(struct flb_s3 *ctx, const char *tag, time_t create_time
         }
     }
+
+    /* Update file and increment index value right before request */
+    if (ctx->key_fmt_has_seq_index) {
```
this feature works for both PutObject and for multipart uploads? Or no?

I think i see that this code has only been added for the PutObject case...

PettitWesley replied:
@StephenLeeY ?

StephenLeeY replied:
Multipart has been added!

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -1524,6 +1678,16 @@ static int cb_s3_exit(void *data, struct flb_config *config)
         }
     }
 
+    if (remove(ctx->seq_index_file) != 0) {
+        flb_plg_error(ctx->ins, "Could not successfully delete metadata file on exit. "
+                      "Overwriting index to 0");
+        ret = write_seq_index(ctx->seq_index_file, 0, 0);
+        if (ret < 0) {
+            flb_plg_error(ctx->ins, "Failed to overwrite index to 0");
+            return -1;
+        }
+    }
```
is overwriting the index file on exit the desired behavior? I thought the whole point of saving the index in a file is so that it could survive restarts. If you delete/overwrite it on shut down then there's no point in having the file I think..

StephenLeeY commented:
Breakdown of changes:
`out_s3`: added sequential indexing feature with index store in filesystem
`aws_util`: added sequential index parsing to flb_get_s3_key
`tests`: added unit tests for aws_util flb_get_s3_key function

Also added fixes based on comments.

StephenLeeY commented:
@PettitWesley

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -252,6 +253,143 @@ static flb_sds_t concat_path(char *p1, char *p2)
+        return -1;
+    }
+
+    ctx->stream_metadata = flb_fstore_stream_create(ctx->fs, "metadata");
```
why `metadata`? Why not something super obvious of its purpose, like `sequence` or `sequence_number`

StephenLeeY replied:
I was keeping in mind possible future changes where we might want to store additional S3 metadata information! I can change this to be more specific to sequential index if that is preferred however.

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -252,6 +253,143 @@ static flb_sds_t concat_path(char *p1, char *p2)
+
+    ctx->key_fmt_has_seq_index = FLB_TRUE;
+
+    /* Initialize local storage to store metadata files */
+    ret = s3_store_init(ctx);
+    if (ret == -1) {
+        flb_plg_error(ctx->ins, "Failed to initialize S3 storage: %s",
+                      ctx->store_dir);
+        return -1;
```
nit: I'd initialize s3 store outside of this function. Because this function is called `init_seq_index` and so I think initializing the entire store should not fall under its ownership

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -252,6 +253,143 @@ static flb_sds_t concat_path(char *p1, char *p2)
+        return -1;
+    }
+    tmp = "/index_metadata";
+    ctx->metadata_dir = flb_sds_cat(ctx->metadata_dir, tmp, strlen(tmp));
```
If you look for `flb_sds_cat` in the code there is an existing pattern where its return value is set to a temporary pointer. Then you check that pointer and then set it to the final pointer. 

This is because under the hood it uses realloc, and on realloc failure, you need to free the original memory: https://stackoverflow.com/questions/1607004/does-realloc-free-the-former-buffer-if-it-fails

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -252,6 +253,143 @@ static flb_sds_t concat_path(char *p1, char *p2)
+
+    ctx->seq_index_file = flb_sds_create(ctx->metadata_dir);
+    if (ctx->seq_index_file == NULL) {
+        flb_plg_error(ctx->ins, "Failed to create sequential index file path");
+        flb_errno();
+        return -1;
```
do you need to free alocated memory before return?

StephenLeeY replied:
No, all memory is accounted for before return.

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -252,6 +253,143 @@ static flb_sds_t concat_path(char *p1, char *p2)
+        flb_errno();
+        return -1;
+    }
+    len = snprintf(NULL, 0, "%d", ctx->ins->id);
+    tmp_buf = flb_malloc(len + 1);
+    if (tmp_buf == NULL) {
+        flb_plg_error(ctx->ins, "Could not allocate enough memory for file name");
+    }
+    snprintf(tmp_buf, len + 1, "%d", ctx->ins->id);
+    tmp_buf[len] = '\0';
+    ctx->seq_index_file = flb_sds_cat(ctx->seq_index_file, tmp_buf, strlen(tmp_buf));
+    flb_free(tmp_buf);
+    if (ctx->seq_index_file == NULL) {
+        flb_plg_error(ctx->ins, "Failed to create sequential index file path");
+        flb_errno();
+        return -1;
```
you don't need to figure out how long the plugin id will be and malloc a temporary buffer, that's inefficient

you can rely on the fact that plugin IDs are capped at something like 256. 

Even if it gets increased, a reasonable max value is may be ~1000, so you can just have char array in the function which you use for this purpose. You never need more than 5 chars.

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -918,6 +1081,10 @@ static int put_all_chunks(struct flb_s3 *ctx)
             continue;
         }
+        /* skip metadata stream */
+        if (fs_stream == ctx->stream_metadata) {
```
should it be `stream_sequence`?

StephenLeeY commented:
@PettitWesley Addressed comments and ready for review.

StephenLeeY commented:
@PettitWesley Addressed comments and ready for review.

PettitWesley commented:
@DrewZhang13 @zhonghui12 I'd like both of you to look at this:
* Bare minimum: At least read through the code once and understand it
* Points will be awarded for: Useful code review that catches issues or helps Stephen improve

zhonghui12 commented on plugins/out_s3/s3.c:
```diff
@@ -252,6 +253,127 @@ static flb_sds_t concat_path(char *p1, char *p2)
+            flb_plg_error(ctx->ins, "Failed to write to sequential index metadata file");
+            return -1;
+        }
+    } else {
```
So in fluent bit, it has its own standard. We should have `else` in a new line: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

zhonghui12 commented on plugins/out_s3/s3.c:
```diff
@@ -1627,8 +1819,8 @@ static struct flb_config_map config_map[] = {
-    "s3_key_format_tag_delimiters option. See the in depth examples and tutorial"
-    " in the documentation."
+    "s3_key_format_tag_delimiters option. Add $INDEX to enable sequential indexing "
+    "for file names. See the in depth examples and tutorial in the documentation."
```
We might need to add one part here that if $INDEX is supported, we will not have random string here.

zhonghui12 commented on src/aws/flb_aws_util.c:
```diff
@@ -796,6 +800,28 @@ flb_sds_t flb_get_s3_key(const char *format, time_t time, const char *tag, char
+        flb_free(seq_index_str);
+    }
+
     /* Find all occurences of $UUID and replace with a random string. */
```
Do we need any `flb_free(seq_index_str )` in this function?

StephenLeeY replied:
If I understand correctly, we should only need one `flb_free` on line 822 because it is malloc-ed on line 806.

StephenLeeY commented:
@PettitWesley Addressed zhonghui and drewzhang's comments!

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -252,6 +253,128 @@ static flb_sds_t concat_path(char *p1, char *p2)
+        flb_errno();
+        return -1;
+    }
+    tmp = "/seq_index_";
```
This should probably actually be a constant defined at the top of the file. Or at the very least it should be defined at the top of the function- fluent bit C code style rules say that all function variables must be defined at the top of the function.

PettitWesley replied:
Oh I see, you declare it at the top, and then set it to different strings as needed. A constant would be prefered, I guess this is fine.

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -1239,6 +1419,19 @@ static struct multipart_upload *create_upload(struct flb_s3 *ctx,
     m_upload->init_time = time(NULL);
     mk_list_add(&m_upload->_head, &ctx->uploads);
 
+    /* Update file and increment index value right before request */
+    if (ctx->key_fmt_has_seq_index) {
+        ctx->seq_index++;
```
If the create multipart upload request fails- will seq_index be decremented? I see the decrement code for put_object, I can't find it for multipart.

StephenLeeY replied:
The `seq_index` decrement is in the multipart section of `upload_data` after `create_multipart_upload` is called. I had to do it this way because error handling for multipart upload requests are handled outside of `create_upload`.

PettitWesley commented on src/aws/flb_aws_util.c:
```diff
@@ -796,6 +800,28 @@ flb_sds_t flb_get_s3_key(const char *format, time_t time, const char *tag, char
 
+    /* Find all occurences of $INDEX and replace with the appropriate index. */
+    if (strstr((char *) format, INDEX_STRING)) {
+        seq_index_len = snprintf(NULL, 0, "%"PRIu64, seq_index);
```
I think there's a typo here- I think you meant `"%PRIu64"`

PettitWesley replied:
Nvm.

PettitWesley commented on src/aws/flb_aws_util.c:
```diff
@@ -796,6 +800,28 @@ flb_sds_t flb_get_s3_key(const char *format, time_t time, const char *tag, char
+            goto error;
+        }
+
+        sprintf(seq_index_str, "%"PRIu64, seq_index);
```
also a typo?

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -252,6 +253,128 @@ static flb_sds_t concat_path(char *p1, char *p2)
+        return -1;
+    }
+
+    ret = fscanf(fp, "%"PRIu64, seq_index);
```
typo or is this is C thing that I don't know?

PettitWesley replied:
Oh I see, it is a C thing that I didn't know about: https://stackoverflow.com/questions/16859500/mmh-who-are-you-priu64

PettitWesley commented:
@StephenLeeY I think this is nearly ready to merge- address my comments, squash the commits to one per component, and then make sure you have thoroughly tested it for both the multipart and put_object case.

StephenLeeY commented:
Ready for merge @PettitWesley

zhonghui12 commented on plugins/out_s3/s3.c:
```diff
@@ -254,6 +255,128 @@ static flb_sds_t concat_path(char *p1, char *p2)
+    }
+
+    /* Create directory path if it doesn't exist */
+    ret = mkdir(ctx->metadata_dir, 0755);
```
I am not sure if we need to use a standard as `0600` here. Maybe @PettitWesley can answer?

PettitWesley replied:
yeah 0600- read/write for the fluent bit user would be preferable. Actually AWS Security has requested that all of these buffer files have that permission

zhonghui12 commented on plugins/out_s3/s3.c:
```diff
@@ -254,6 +255,128 @@ static flb_sds_t concat_path(char *p1, char *p2)
+        ctx->seq_index = 0;
+        ret = write_seq_index(ctx->seq_index_file, ctx->seq_index);
+        if (ret < 0) {
+            flb_plg_error(ctx->ins, "Failed to write to sequential index metadata file");
```
Do we need to do `flb_errno();` here?

StephenLeeY replied:
We would be failing on a `fscanf`, `fprintf`, or `fopen` call. Not too sure if that means we need `flb_errno()`?

PettitWesley replied:
`fopen` and `fscanf` both set `errno`. However, the call to `flb_errno()` should happen inside the `write_seq_index` function directly after calling those functions that might set errno. Not out here.
